TL124A_MemMtrc/doc/MemMtrc.1.ronn
MemMtrc(1) -- Nexteer Embedded Software Memory Metrics Tool
========================================================================================================================

## SYNOPSIS ##
MemMtrc.py [-h] [-v] [-V] [-d DIR] [-e ELF] [-f FEE] [-l LINKER] [-o OUTPUT] [-t {text,html}]



## DESCRIPTION ##
This tool was designed to extract information regarding memory utilization metrics from EA4 software projects and
generate documentation detailing the utilization.  The tool automatically parses information from a series of
configuration and output files in an integration project to produce an output report.


## ARGUMENTS ##
The memory metrics tool does not take any positional arguments.  For information on the optional arguments, see below.



## OPTIONS ##
* `-h`, `--help`:
    Print a help message and exit.

* `-v`, `--verbose`:
    Output detailed information to console.  This option can be doubled (i.e. `-vv` or `--verbose --verbose`) to emit
    even more information.  Be aware that all commands issued to Synergy CCM will be printed to the console when this
    argument is doubled.  This includes the `start` command which includes a plain-text copy of your password.

* `-V`, `--version`:
    Display version information and exit.

* `-d DIR`, `--dir DIR`:
    The path to the integration prject.  If omitted, the tool will search '../../' for a subdirectory containing
    *_EPS_* in the name.  The path will be used to search for the input files listed below according to the standard
    project structure.  If absolute pathes to each of the inputs are supplied using optional arguments for `--fee`,
    `--linker`, and `--elf` then this argument may be omitted as it will be ignored by the tool.

* `-e ELF`, `--elf ELF`:
    Specify the absolute path to the .elf file generated when the project is compiled.  This file will be parsed for
    information pertaining to the size of sections as well as the software name and version number.  If omitted, the
    tool will search within the output directory in the integration project.  See `--dir` for more information about how
    the integration project path is specified or derrived.

* `-f FEE`, `--fee FEE`:
    Specify the path to the FEE configuration file.  The tool parses this file to gather information regarding the FEE
    block configuration and usage for the project.  If omitted, the autosar/Config/ECUC directory will be searched
    within the integration project for a file named EPS_Fee_ecuc.arxml.  See `--dir` for more information about how
    the integration project path is specified or derrived.

* `-l LINKER`, `--linker LINKER`:
    Specify the path to the linker command file for the project.  This file is used to parse memory sections to
    understand how Flash memory has been divided and to provide information about the utilization of each of the
    sections.  This tool expects specific formatting of part of this file to work properly, see the
    [LINKER COMMAND FILE FORMAT][] section below for additional information.  If omitted, the tool will search within
    the tools directory of the integration project for a single file ending with a *.ld extension.  Further, the
    filename of the linker command file is expected to be the model number of the processor for the project.  This
    information is used to select the proper Processor class within the tool which defines the memory ranges that are
    available for the program..  See `--dir` for more information about how the integration project path is specified
    or derrived.

* `-o OUTPUT`, `--output OUTPUT`:
    Report output path and/or filename.  Defaults to ../output/MemoryMetrics_<Timestamp> if omitted.  The output
    directory will be created if it does not exist.

* `-t {text,html}`, `--type {text,html}`:
    Report output format.  The filename extension will automatically be updated if incorrect to apply the proper
    extension for the selected format.  The text format does not produce a file as output but rather prints the results
    to the terminal through stdout.



## RETURN VALUES ##
* `1`:
    Either the path specified by `--dir` is invalid or the tool was unable to locate the integration project directory
    in the folder up two directories from the tool script.  If `--dir` is omitted, the tool will search the directory
    that is located two directories up from `MemMtrc.py` for a directory containing *_EPS_* in the name.  If not found
    this return code will be given.  If the integration project does not reside in a folder with EPS in the name, then
    the `--dir` switch must be used to manually specify the integration project folder path.

* `2`:
    Either the .elf file provided using the `--elf` switch is invalid or the tool was unable to automatically locate
    an .elf file in the output folder of the integration project.  This can happen if the project has never been built.

* `3`:
    Either the .arxml file specified for the `--fee` switch is invalid or the tool was unable to find the file in the
    ECUC config file directory (see `--fee` in the [OPTIONS][] section for more information about the default, expected
    filename and path).

* `4`:
    Either the linker command file specified with the `--linker` switch is invalid or the tool was unable to find a
    linker command file (file with a .ld extension) in the tools folder of the integration project.

* `5`:
    The tool attempted to find the linker command file in the tools folder of the integration project automatically but
    found more than one file with a .ld extension.  To respolve this issue, use the `--linker` switch so manually
    specify the linker command file that should be used.

* `6`:
    Invalid output filename provided for the `--output` switch.  Ensure that the filename is provided or omit the
    `--output` option.

* `7`:
    Invalid report format type specified.  The only supported report output formats are text and html.

* `8`:
    The processor type, as determined from the linker command file name, is unknown by this tool.  An update to this
    tool to define the processor type is required.  Without this change the tool is not aware of the memory types or
    ranges for the processor.



## LINKER COMMAND FILE FORMAT ##
The linker command file must follow a specific format for the memory metrics tool to properly parse information from the
file.  The MEMORY section within the file must be split into sections separated by two newline characters.  The sctions
may also start with a comment (/&ast; ... &ast;/) containing a textual description of the section.  Below is an example
configuration taken from the T1xx GA project:

    MEMORY
    {
        /* Bootloader 0x0000 0000 - 0x0000 17FFF */

        /* Application */
        GMAPPHEADER      (RX) : ORIGIN=0x00018000, LENGTH=0x00000200
        APPVECTOR        (RX) : ORIGIN=0x00018200, LENGTH=0x00000800
        APP_FLASH        (RX) : ORIGIN=0x00018A00, LENGTH=0x0009F3F0
        APP_CRC          (RX) : ORIGIN=0x000B7DF0, LENGTH=0x00000010
        APPPRESENCE      (RX) : ORIGIN=0x000B7E00, LENGTH=0x00000100
        APP_VALDBLK      (RX) : ORIGIN=0x000B7F00, LENGTH=0x00000100

        /* Calibration 2 - Feature Cals */
        GMCAL2HEADER     (RX) : ORIGIN=0x000B8000, LENGTH=0x00000020
        CAL2_FLASH       (RX) : ORIGIN=0x000B8020, LENGTH=0x00007DD0
        CAL2_CRC         (RX) : ORIGIN=0x000BFDF0, LENGTH=0x00000010
        CAL2PRESENCE     (RX) : ORIGIN=0x000BFE00, LENGTH=0x00000100
        CAL2_VALDBLK     (RX) : ORIGIN=0x000BFF00, LENGTH=0x00000100

        /* Calibration 1 - System Cals */
        GMCAL1HEADER     (RX) : ORIGIN=0x000C0000, LENGTH=0x00000020
        CAL1_FLASH       (RX) : ORIGIN=0x000C0020, LENGTH=0x0001FDD0
        CAL1_CRC         (RX) : ORIGIN=0x000DFDF0, LENGTH=0x00000010
        CAL1PRESENCE     (RX) : ORIGIN=0x000DFE00, LENGTH=0x00000100
        CAL1_VALDBLK     (RX) : ORIGIN=0x000DFF00, LENGTH=0x00000100

        /* Calibration 3 - Performance Cals */
        GMCAL3HEADER     (RX) : ORIGIN=0x000E0000, LENGTH=0x00000020
        CAL3_FLASH       (RX) : ORIGIN=0x000E0020, LENGTH=0x0001FDD0
        CAL3_CRC         (RX) : ORIGIN=0x000FFDF0, LENGTH=0x00000010
        CAL3PRESENCE     (RX) : ORIGIN=0x000FFE00, LENGTH=0x00000100
        CAL3_VALDBLK     (RX) : ORIGIN=0x000FFF00, LENGTH=0x00000100

        iROM_1         : ORIGIN = 0x01000000,  LENGTH = 32k
        iRAM_0         : ORIGIN = 0xFEBE0000,  LENGTH = 124k
        iRAM_DMAWRITE  : ORIGIN = 0xFEBFF000,  LENGTH = 4k
    }

Important notes from this configuration:

1. The tool will see the bootloader comment as a "section" but becuase it contains no ORIGIN or LENGTH it will be
   ignored.
2. The iRxM entries at the bottom of the MEMORY section reside outside of the Flash memory region for this processor
   and will therefore be ignored.
3. This configuration would produce 4 sections in the resultant report, one each for the Application, and three
   calibration regions.  The sections would be labelled "Application", "Calibration 2 - Feature Cals", "Calibration 1 -
   System Cals", and "Calibration 3 - Performance Cals", respectively in the report.
4. It is critical that the newlines between sections are back-to-back to ensure the proper split.  This means that the
   lines between sections must **not** contain any whitespace characters such as tabs or spaces.



## BUGS ##
Bugs should be reported and tracked using the Anomaly system in [Change Synergy](http://10.199.100.212:8603/change).
The use of a "Continuous Improvement" CR is recommended in favor of writing an anomaly against a particular program.



## HISTORY ##
- `01.00.00 - 2017-12-31`
    - First release of tool.
    - Runs against EA4 programs to generate Flash, RAM, and FEE usage information.
    - Works with both Fee and SmallSector Fee.


## DOCUMENTATION ##
This documentation was generated using [ronn](http://rtomayko.github.io/ronn).  `ronn` converts a text file with a
format similar to Markdown into an HTML file with a format typical of a standard manpage.  For more information about
`ronn` see it's manpage: [ronn(1)](http://rtomayko.github.io/ronn/ronn.1.html).  For more information on the syntax see
that manpage: [ronn-format(7)](http://rtomayko.github.io/ronn/ronn-format.7.html).



## AUTHOR ##
Jared Julien <jared.julien@nexteer.com>



## COPYRIGHT ##
(c) 2017 Nexteer Automotive


TL124A_MemMtrc/src/MemMtrc.py
"""=====================================================================================================================
#      File: Ea4MemoryMetrics.py
#   Project: EA4 project memory metrics parsing tool
#   Created: 2017-12-20
#    Author: Jared Julien
#     Email: jared.julien@nexteer.com
# Copyright: (c) 2017 Nexteer Automotive
#
# Date        Rev   Author    Change Description
# ==========  ====  ========  =========================================================================================
# 2017-12-20  1     Jared     Initial file creation
# -------------------------------------------------------------------------------------------------------------------"""

# ======================================================================================================================
# Import Statements
# ----------------------------------------------------------------------------------------------------------------------
# Python Standard Libraries
import os
import sys
import glob
import argparse
import logging
import datetime

# Add the include directory for this component to the system path to allow importing of local libraries.
relInclude = os.path.join('..', 'include')
absInclude = os.path.abspath(relInclude)
sys.path.insert(1, absInclude)

# 3rd Party Libraries
import tabulate  # https://pypi.python.org/pypi/tabulate/0.7.7

# 1st Party Libraries
import renesas
import report


# ======================================================================================================================
# Constants
# ----------------------------------------------------------------------------------------------------------------------
DIE_NO_INTEGRATION_PROJECT = 1
DIE_NO_ELF = 2
DIE_NO_FEE_ARXML = 3
DIE_NO_LINKER = 4
DIE_TOO_MANY_LINKER = 5
DIE_INVALID_OUTPUT = 6
DIE_INVALID_TYPE = 7
DIE_PROCESSOR_NOT_SUPPORTED = 8


# ======================================================================================================================
# Software Version Number
# ----------------------------------------------------------------------------------------------------------------------
VERSION_MAJOR = 1
VERSION_MINOR = 0
VERSION_BUILD = 0
VERSION = "%02d.%02d.%02d" % (VERSION_MAJOR, VERSION_MINOR, VERSION_BUILD)


# ======================================================================================================================
# Helper Functions
# ----------------------------------------------------------------------------------------------------------------------
def print_summary(device):
    """Print to the console a summay of the memory regions for the supplied processor."""
    table = []
    for flash in device.flash:
        print
        print flash.name
        for section in flash.sections:
            table.append([section.name, hex(section.start), section.length])
        print tabulate.tabulate(table)
    for ram in device.ram:
        print
        print ram.name
        for section in ram.sections:
            table.append([section.name, hex(section.start), section.length])
        print tabulate.tabulate(table)
    for feeMem in device.fee:
        for region in feeMem.regions:
            print region
            for block in region.blocks:
                print ' ', block
    for mem in device.memories:
        print '%s: %0.2f%%' % (mem.name, mem.usage * 100)


# ======================================================================================================================
# Main Function
# ----------------------------------------------------------------------------------------------------------------------
if __name__ == '__main__':
    description = "Nexteer Embedded Software Metrics Tool"
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument(
        '-v', '--verbose',
        default=0,
        action='count',
        help='Output detailed information to console'
    )
    parser.add_argument(
        '-V', '--version',
        action='version',
        version='%%(prog)s %s' % VERSION,
        help='Display version number and exit'
    )
    parser.add_argument(
        '-d', '--dir',
        help='Integration project directory used to attempt auto-location of input files'
    )
    parser.add_argument(
        '-e', '--elf',
        help='Override path to ELF file (If omitted, tool will search within dir above)'
    )
    parser.add_argument(
        '-f', '--fee',
        help='Override path to FEE ECUC ARXML file (If omitted, tool will search within dir above)'
    )
    parser.add_argument(
        '-l', '--linker',
        help='Override path to linker command file (If omitted, tool will search within dir above)'
    )
    parser.add_argument(
        '-o', '--output',
        help='Filename of the report output (defaults to ../output/MemoryMetrics_<Timestamp>)'
    )
    parser.add_argument(
        '-t', '--type',
        choices=['text', 'html'],
        default='html',
        help='Report output format'
    )
    args = parser.parse_args()

    # Setup logging
    levels = [logging.WARNING, logging.INFO, logging.DEBUG]
    logLevel = levels[min(args.verbose, 2)]
    logging.basicConfig(level=logLevel, format='%(levelname)s: %(message)s')

    # Determine the path to the integration project.
    if args.dir:
        integration = args.dir
        if not os.path.exists(integration):
            logging.critical('Specified integration project path is invalid')
            sys.exit(DIE_NO_INTEGRATION_PROJECT)
    elif not args.elf or not args.fee or not args.linker:
        # If the inputs dictate pathes to all of the required files, don't bother looking for the integration project.
        path = os.path.join('..', '..', '*_EPS_*')
        matches = glob.glob(path)
        if len(matches) != 1:
            logging.critical('Cannot determine path to integration project within specified directory')
            sys.exit(DIE_NO_INTEGRATION_PROJECT)
        integration = matches[0]

    # Determine path to ELF file.
    if args.elf:
        # If specified, use that file.
        elf = args.elf
        # Ensure that the input file exits.
        if not os.path.exists(elf):
            logging.critical('Specified ELF file does not exist')
            sys.exit(DIE_NO_ELF)
    else:
        # Search the integration project for the .elf file in the typical output folder location.
        path = os.path.join(integration, 'output', '*.elf')
        logging.info('No path to ELF supplied as an input, automatically searching %s for one', path)
        matches = glob.glob(path)
        if not matches:
            # No .elf file in output folder.
            logging.critical('No elf file found in output path of integration project - has this project been built?')
            sys.exit(DIE_NO_ELF)
        elif len(matches) == 1:
            # Exactly one .elf file in output folder (best case scenario).
            elf = matches[0]
            short = os.path.basename(elf)
            logging.info('Found only one ELF file in output directory, using that (%s)', short)
        else:
            # More than one .elf file in output folder, lets use the newest one.
            # Sort the matches using the creation time in reverse order.
            matches.sort(key=os.path.getctime, reverse=True)
            # The above sort ensures that the zeroth entry in the list is now the newest .elf file.
            elf = matches[0]
            short = os.path.basename(elf)
            logging.warning('More than one ELF file found in output directory, using newest (%s)', short)
            logging.warning('If this is incorrect use the --elf switch to specify the proper elf manually')

    # Determine path to FEE ECUC ARXML file.
    if args.fee:
        # If specified, use that file.
        fee = args.fee
        # Ensure that the specified FEE file exists.
        if not os.path.exists(fee):
            logging.critical('Specified FEE ECUC ARXML file does not exist')
            sys.exit(DIE_NO_FEE_ARXML)
    else:
        # Search the integration project for the FEE ARXML at the usual location in the autosar folder.
        logging.info('No path to FEE ECUC ARXML supplied as an input')
        fee = os.path.join(integration, 'autosar', 'Config', 'ECUC', 'EPS_Fee_ecuc.arxml')
        if not os.path.exists(fee):
            logging.critical('Unable to locate FEE ECUC file at %s as expected', fee)
            logging.critical('Specify an ECUC file manually using --fee switch')
            sys.exit(DIE_NO_FEE_ARXML)

    # Determine path to the linker command file.
    if args.linker:
        # If a specific path was give, use that file.
        linker = args.linker
        # Ensure that the specified linker command file exists.
        if not os.path.exists(linker):
            logging.critical('Specified linker command file does not exist')
            sys.exit(DIE_NO_LINKER)
    else:
        # Search for the linker command file in the tools directory of the integration project.
        path = os.path.join(integration, 'tools', '*.ld')
        logging.info('No path to linker command file supplied as an input, automatically searching %s for one', path)
        matches = glob.glob(path)
        if not matches:
            # No .ld file in output folder.
            logging.critical('No linker command file found in tools path of integration project')
            sys.exit(DIE_NO_LINKER)
        elif len(matches) == 1:
            # Exactly one .ld file in output folder (best case scenario).
            linker = matches[0]
            short = os.path.basename(linker)
            logging.info('Found only one ELF file in output directory, using that (%s)', short)
        else:
            # More than one linker command file?!? Not sure which one to use.
            logging.critical('More than one linker command file was found - unable to determine which to use.')
            logging.critical('Use the --linker switch to specify the correct file')
            sys.exit(DIE_TOO_MANY_LINKER)

    # Determine the processor type from the linker command file name.
    linkerFile = os.path.basename(linker)
    processorName = os.path.splitext(linkerFile)[0]
    try:
        processorType = getattr(renesas, processorName)
    except AttributeError:
        logging.critical('Processor used by this project is not supported by the current version of this tool')
        sys.exit(DIE_PROCESSOR_NOT_SUPPORTED)

    # Print a debug report showing all of the pathes.
    logging.debug('ELF: %s', elf)
    logging.debug('Fee: %s', fee)
    logging.debug('Linker: %s', linker)

    # Build up a processor object from the supplied files.
    processor = processorType(elf, fee, linker)

    # Determine output filename.
    if args.output:
        output = args.output
        if not os.path.splitext(output)[0]:
            logging.critical('Specified output filename is invalid')
            sys.exit(DIE_INVALID_OUTPUT)
    else:
        timestamp = datetime.datetime.now().strftime('%Y%m%dT%H%M')
        output = os.path.join('..', 'output', 'MemoryMetrics_%s' % timestamp)

    # Create the output directory if it doesn't exist.
    folder = os.path.dirname(output)
    if not os.path.exists(folder):
        os.makedirs(folder)

    # Extract the extension to be used at generation to ensure application of proper extension.
    base = os.path.basename(output)
    filename, extension = os.path.splitext(base)

    # Generate the report.
    if args.type == 'text':
        print_summary(processor)
    elif args.type == 'html':
        if extension not in ['.htm', '.html']:
            output += '.html'
        report.html.generate(processor, output)
    else:
        logging.critical('Invalid report type specified')
        sys.exit(DIE_INVALID_TYPE)


# End of File
